### ISSUE TITLE
[Clang] -Winvalid-constexpr misbehaves in C++23 mode

### ISSUE THREAD

**Issue by frederick-vs-ja** [llvm/llvm-project#149041]
In C++23/26 mode, Clang rejects the following function definition with `-Winvalid-constexpr` used, saying "constexpr function never produces a constant expression". [Godbolt link](https://godbolt.org/z/WaKPMjorP).
```C++
constexpr bool fun(const bool& b) { return b; }
```

This seems related to [P2280R4](https://wg21.link/p2280r4), but the incorrect diagnostic occurs only since Clang 21.

**Comment by cor3ntin** [comment]
@shafik @efriedma-quic

**Comment by efriedma-quic** [comment]
-Winvalid-constexpr, formally speaking, shouldn't be a pedantic error in C++23: the standard allows "invalid" constexpr functions.  The implementation of the flag is weird.  We should have two separate diagnostics: the default-error ExtWarn, and the off-by-default C++20-compat warning, and let the diagnostic system do the rest.

That said, there are other reasons to get this right.  Probably we just need to suppress the "read of variable 'b' whose value is not known" diagnostic when we're checking if an expression is potentially-constexpr.
